pandas/tests/indexes/test_setops.py

```diff
@@ -484,3 +485,78 @@ def check_intersection_commutative(left, right):
 
     check_intersection_commutative(idx, idx_non_unique)
     assert idx.intersection(idx_non_unique).is_unique
+
+
+@pytest.mark.parametrize(
+    "cls",
+    [
+        Int64Index,
+        Float64Index,
+        DatetimeIndex,
+        CategoricalIndex,
+        TimedeltaIndex,
+        lambda x: Index(x, dtype=object),
+    ],
+)
+def test_union_duplicate_index_subsets_of_each_other(cls):
+    # GH#31326
+    a = cls([1, 2, 2, 3])
+    b = cls([3, 3, 4])
+    expected = cls([1, 2, 2, 3, 3, 4])
+    if cls is CategoricalIndex:
```

Comment:
these are strange for b.union(a, sort=False)

Reply:
We are only calculating a.union(b) with sort=None and sort=False. b.union(a) would look differently of course. Should I add them?